osu.Game/Rulesets/Mods/ModFlashlight.cs

```diff
@@ -1,17 +1,158 @@
         public override ModType Type => ModType.DifficultyIncrease;
         public override string Description => "Restricted view area.";
         public override bool Ranked => true;
+
+        public const double FLASHLIGHT_FADE_DURATION = 800;
+        protected readonly BindableInt Combo = new BindableInt();
+
+        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            Combo.BindTo(scoreProcessor.Combo);
+        }
+
+        public virtual void ApplyToRulesetContainer(RulesetContainer<T> rulesetContainer)
+        {
+            var flashlight = CreateFlashlight();
+            flashlight.Combo = Combo;
+            flashlight.RelativeSizeAxes = Axes.Both;
+            flashlight.Colour = Color4.Black;
+            rulesetContainer.KeyBindingInputManager.Add(flashlight);
+
+            flashlight.Breaks = rulesetContainer.Beatmap.Breaks;
+        }
+
+        public abstract Flashlight CreateFlashlight();
+
+        public abstract class Flashlight : Drawable
+        {
+            internal BindableInt Combo;
+            private Shader shader;
+            protected readonly FlashlightUniformWrapper MousePosWrapper = new FlashlightUniformWrapper();
+
+            protected override DrawNode CreateDrawNode() => new FlashlightDrawNode();
+
+            public override bool RemoveCompletedTransforms => false;
+
+            public List<BreakPeriod> Breaks;
+
+            protected override void ApplyDrawNode(DrawNode node)
+            {
+                base.ApplyDrawNode(node);
+
+                var flashNode = (FlashlightDrawNode)node;
+
+                flashNode.Shader = shader;
+                flashNode.ScreenSpaceDrawQuad = ScreenSpaceDrawQuad;
+                flashNode.MousePosWrapper = MousePosWrapper;
+            }
+
+            [BackgroundDependencyLoader]
+            private void load(ShaderManager shaderManager)
+            {
+                shader = shaderManager.Load("PositionAndColour", "Flashlight");
+            }
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                Combo.ValueChanged += OnComboChange;
+
+                this.FadeInFromZero(FLASHLIGHT_FADE_DURATION);
+
+                foreach (var breakPeriod in Breaks)
+                {
+                    this.Delay(breakPeriod.StartTime + FLASHLIGHT_FADE_DURATION).FadeOutFromOne(FLASHLIGHT_FADE_DURATION);
+                    this.Delay(breakPeriod.EndTime - FLASHLIGHT_FADE_DURATION).FadeInFromZero(FLASHLIGHT_FADE_DURATION);
+                }
+            }
+
+            protected abstract void OnComboChange(int newCombo);
+        }
+
+        public class FlashlightUniformWrapper
+        {
+            public bool Rectangular;
+            public bool RectangularChanged = true;
+
+            public Vector2 FlashlightPosition;
+            public bool FlashlightPositionChanged = true;
+
+            public float CircularFlashlightSize;
```

Comment:
Like I mentioned in the other comment, I really don't think we need this separation of size, position, and the rectangular thing you have here. They should be re-usable for all flashlight implementations.